include/nuttx/can/can.h

```diff
@@ -491,7 +491,13 @@ begin_packed_struct struct can_msg_s
 
 struct can_rxfifo_s
 {
-  sem_t         rx_sem;                  /* Counting semaphore */
+  /* Binary semaphore. Indicates whether FIFO is available for reading
+   * AND not empty. Only take this sem inside a critical section to guarantee
+   * exclusive access to both the semaphore and the head/tail FIFO indices.
+   */
+
+  sem_t         rx_sem;
```

Comment:
but, the code still use it as the lock, not an indicator?

Reply:
No; `rx_sem` is not used as a lock because the ISR (which calls `can_receive()`) can edit the FIFO whenever it wants as long as interrupts are enabled. `rx_sem` exists to allow `can_receive()` to wake up a reader thread, but the reader thread then has exclusive access only because it uses a critical section, not because it took the semaphore.

See also my reply above.

Actually, `rx_sem` was a binary semaphore before too, but "0" and "1" didn't have any defined meaning because it was initialized to 1 and then immediately decremented in a loop in `can_read()` and `can_poll()`